src/main/scheduler/scheduler.c

```diff
@@ -564,36 +615,45 @@ FAST_CODE void scheduler(void)
             taskRequiredTimeCycles += taskGuardCycles;
 
             if (!gyroEnabled || (taskRequiredTimeCycles < schedLoopRemainingCycles)) {
+                uint32_t antipatedEndCycles = nowCycles + taskRequiredTimeCycles;
                 taskExecutionTimeUs += schedulerExecuteTask(selectedTask, currentTimeUs);
                 nowCycles = getCycleCounter();
-                schedLoopRemainingCycles = cmpTimeCycles(nextTargetCycles, nowCycles);
+                int32_t cyclesOverdue = cmpTimeCycles(nowCycles, antipatedEndCycles);
+
+#if defined(USE_LATE_TASK_STATISTICS)
+                if (cyclesOverdue > 0) {
+                    if ((currentTask - tasks) != TASK_SERIAL) {
+                        DEBUG_SET(DEBUG_SCHEDULER_DETERMINISM, 1, currentTask - tasks);
+                        DEBUG_SET(DEBUG_SCHEDULER_DETERMINISM, 2, clockCyclesTo10thMicros(cyclesOverdue));
+                        currentTask->lateCount++;
+                        lateTaskCount++;
+                        lateTaskTotal += cyclesOverdue;
+                    }
+                }
+#endif  // USE_LATE_TASK_STATISTICS
 
-                if (schedLoopRemainingCycles < taskGuardMinCycles) {
+                if ((cyclesOverdue > 0) || (-cyclesOverdue < taskGuardMinCycles)) {
```

Comment:
?

Reply:
We aim to have tasks complete early by at least `taskGuardMinCycles`. We therefore allow an extra amount of time, `taskGuardCycles` for a task to execute, only calling it if the number of remaining cycles in the loop is greater than or equal to `taskRequiredTimeCycles` plus `taskGuardCycles`. `taskGuardCycles` is constrained to be  between `taskGuardMinCycles` and `taskGuardMaxCycles`. The above is ensuring that if we're actually late, or early by less than `taskGuardMinCycles`, then `taskGuardCycles` is increased.